### ISSUE TITLE
SSP: Constraint on water table height

### ISSUE THREAD

**Issue by smiths** [JacquesCarette/Drasil#1312]
I don't think we have the ability to do it yet, but there is a missing data constraint in SSP.  The water table height, when it is above the soil, has to be constant.  (The water above the soil is going to find a constant level; it isn't going to have a slope to it.)  As the input is right now, the water surface could follow any function.  We can't do this right now, because, as far as I can tell, the water table height isn't even listed as one of the inputs for SSP in Drasil.  (It is listed in the manual version, kind of.)  This is also a relatively complex constraint, since it isn't simply bounding a value, as in the existing cases.  As an added challenge, it also only applies when the water is above the soil surface, and then only for the distance that it is continuously above the surface.  It is technically possible to have two (or more) constant heights for the water above the soil, if the water table dips below the soil layer and then "pops" up again.

@szymczdm, this is an example of a data constraint that might be interesting for your paper.  People that develop scientific software don't generally think this deeply about what constraints exist for valid data.  The Drasil approach creates a reason to think more intently about what it means for input data to be correct.  (We've talked on separate occasions about the importance of stating what it means for output to be correct.)

**Comment by bmaclach** [comment]
Mathematically, I think this constraint can be represented by:
∀ _i_ : ℕ | _i_ < |___y<sub>wt</sub>___|-1 ∧ (_y<sub>wt,i</sub>_ > _y<sub>slope,i</sub>_ ∨ _y<sub>wt,i+1</sub>_ > _y<sub>slope,i+1</sub>_) : _y<sub>wt,i</sub>_ = _y<sub>wt,i+1</sub>_

You are correct that this is currently not possible in Drasil. Water table height is listed as an input, however, in Drasil (see the "Required Inputs" table in the Functional Requirements section, or the "Inputs" fields of the IMs).

**Comment by JacquesCarette** [comment]
Mathematically, that is a fine formula.  Knowledge-encoding-wise, it is problematic, i.e. it will be hard to "do" anything with it other than to write it down.  What, to me, it really says is:
- for all valid slices (i.e. i is in range),
  piecewise( y_wt > y_slope of this slice, then y_wt this slice = y_wt next slice
                  y_wt > y_slope of next slice, then y_wt this slice = y_wt next slide
                  otherwise, no information gained)

And, now that I read @smiths original issue, I see that his phrasing is even better "the water table height, when it is above the soil, has to be constant". That is the true constraint that has to be 'encoded'. So I think that piecewise is over-complicated.

I think it is just
- exists c. for all valid slices, y_wt,i > y_slope,i, y_wt,i = c

**Comment by bmaclach** [comment]
As @smiths said in the original issue, it's possible to have two or more different constants if the water table rises above the slope in multiple different places. Also, with that formula, the slices where the water table intersects (y_wt = y_slope) can have a y_wt other than c, which shouldn't be possible. Replacing the ">" with a "≥" doesn't fix the issue because then cases where the water table follows the slope exactly over multiple slices are not possible. So it does need to be a little more complicated. Something like:

- for all adjacent groups of valid slices, exists c. for all valid slices in the group, y_wt > y_slope of this slice or next slice, y_wt of this slice = c

**Comment by JacquesCarette** [comment]
Indeed. I am not surprised that the actual formulation needs to be more complicated than that.

"adjacent group of valid slices" sounds like you need a partition of the interval into connected components, and then things a component-wise constant when above the water table.

**Comment by smiths** [comment]
Thank you @bmaclach and @JacquesCarette.  This is a great conversation.  I really believe that the "Drasil process" is forcing us to ask questions that are nearly always ignored in scientific software.  Another solution, for a given SSP "family member" would be to simplify the problem scope.  If we added the assumption that the water table will never be above the soil surface, all of this goes away.  Alternatively, we could have an assumption that there will be at most one continuous region where the water table is above the soil slope; that is, the possibility of the water table popping up and then down and then up again would be excluded.  In all of the problems we dealt with when I took soil mechanics, this was the case.  For practical purposes, I don't think this assumption would rule out the normal engineering situations.  To really know, we would have to consult the "client," who would be someone like Brandon (the original author of the Matlab version of ssp).  As it is right now, I believe his software could be given "bad" input and it would proceed to a solution.  I don't think he worried about this because for typical engineering problems, this won't come up.  

However, we might want to look at the case of a water table above the soil surface with a non-constant height.  I believe Brandon's code would allow a water surface that had peaks and valleys, and is above the soil surface.  If we can document that this shouldn't be possible, and check our input to prevent this case getting through, we have yet another example for @szymczdm and his paper.  :smile:

**Comment by bmaclach** [comment]
Note for this that there are a number of open issues about expanding our constraint language, including some issues for expressing constraints on elements of a list/vector (#1796, #1797, #1799). Once those are complete, we should have a better idea of how to express constraints like this one.

**Comment by smiths** [comment]
Thanks for collecting the issues @bmaclach.
